0015-estimator-interface.md

```diff
@@ -0,0 +1,502 @@
+>> EstimatorResult(
+>>     TaskResult<{evs: ndarray<4, 32>, stds: ndarray<4, 32>}, local_metadata>, 
+>>     TaskResult<{evs: ndarray<18>, stds: ndarray<18>}, local_metadata>, 
+>>     global_metadata
+>> )
+```
+
+The above example shows various features of the design, including combining large numbers of observables with large numbers of parameter value sets for a single circuit.
+
+The detailed targetted signature, following the [deprecation period](#migration-path),
+is as follows:
+
+```python
+T = TypeVar("T", bound=Job)
+
+BindingsArrayLike = Union[
+    BindingsArray, 
+    NDArray, 
+    Dict[Parameter | Tuple[Parameter], NDArray]
+]
+
+ObservablesArrayLike = Union[
+    ObservablesArray,
+    Iterable[str | Pauli | SparsePauliOp]
+]
+
+EstimatorTaskLike = Union[
+    EstimatorTask,
+    Tuple[QuantumCircuit, ObservablesArrayLike, BindingsArrayLike]
+]
+
+class EstimatorBase(ABC, Generic[T]):
+    ...
+    def run(self, tasks: EstimatorTaskLike | Iterable[EstimatorTaskLike], **options) -> T:
+        pass
+```
+
+### Migration Examples <a name="migration-examples"></a>
+
+Technical details of migration are discussed in the [Migration Path](#migration-path) section, but it is perhaps worthwhile to show some examples that demonstrate that, despite the new array and broadcasting abilities, the interface is not overly complicated for simple examples.
+
+One circuit, many observables:
+
+```python
+    # old API
+    estimator.run([circuit] * 4, observables=["ZIII", "IZII", "IIZI", "IIIZ"]).result()
+    >> EstimatorResult(np.array([0.1, 0.2, 0.3, 0.4]), metadata={...})
+
+    # new API (we can use a single task)
+    estimator.run((circuit, ["ZIII", "IZII", "IIZI", "IIIZ"])).result()
+    >> [ResultTask({"evs": np.array([0.1, 0.2, 0.3, 0.4]), "stds": <...>}, metadata={...})]
+```
+
+One circuit, one observable, many parameter value sets:
+
+```python
+    # old API
+    estimator.run([circuit] * 3, observables=["XYZI"] * 3, parameter_values=[[1,2,3], [3,4,5], [8,7,6]]).result()
+    >> EstimatorResult(np.array([0.1, 0.2, 0.3]), metadata={...})
+
+    # new API (we can use a single task)
+    estimator.run((circuit, ["XYZI"], [[1,2,3], [3,4,5], [8,7,6]])).result()
+    >> [ResultTask({"evs": np.array([0.1, 0.2, 0.3]), "stds": <...>}, metadata={...})]
+```
+
+Three different circuits, three different observables:
+
+```python
+    # old API
+    estimator.run([circuit1, circuit2, circuit3], observables=["ZIII", "IZII", "IIZI"]).result()
+    >> EstimatorResult(np.array([0.1, 0.2, 0.3]), metadata={...})
+
+    # new API (unique circuits always need their own tasks)
+    estimator.run([(circuit1, "ZIII"), (circuit2, "IZII"), (circuit3, "IIZI")]).result()
+    >> [
+    >>     ResultTask({"evs": np.array(0.1), "stds": <...>}, metadata={...}), 
+    >>     ResultTask({"evs": np.array(0.2), "stds": <...>}, metadata={...}), 
+    >>     ResultTask({"evs": np.array(0.3), "stds": <...>}, metadata={...})
+    >> ]
+```
+
```

Comment:
In one of the examples, you can see the run method being invoked as `estimator.run((circuit, ["XYZI"], [[1,2,3], [3,4,5], [8,7,6]]))`, which is already a coersion from what should more properly `estimator.run([(circuit, ["XYZI"], [[1,2,3], [3,4,5], [8,7,6]])])` aimed at reducing the number of parens that need to be typed.

However, it seems like you make be asking for even one more: estimator.run(circuit, ["XYZI"], [[1,2,3], [3,4,5], [8,7,6]])`. In my opinion, this would be a step too far, as it would fundamentally change the Python signature of the method.

Reply:
It would change the signature, the same way it did with the original interface's signature, and maybe this use-case is not a priority anymore. But I wanted to make sure you were aware that this actually the most common use of primitives in most high level applications (which are based on variational algorithms, such as VQE) and that this was feedback that we received and ended up being acted on in the past.

Reply:
Thanks for emphasizing this point. We'll do our best to have strong examples in this regard.